polymorphia-backend/src/main/java/com/agh/polymorphia_backend/service/EquipmentService.java

```diff
@@ -0,0 +1,127 @@
+    private final ItemRepository itemRepository;
+
+    public List<EquipmentItemResponseDto> getEquipmentItems(Long courseId) {
+        Long animalId = animalService.validateAndGetAnimalId(courseId);
+        List<AssignedItem> assignedItems = assignedRewardService.getAnimalAssignedItems(animalId);
+        List<Long> assignedItemIds = getAssignedRewardsIds(new ArrayList<>(assignedItems));
+        List<Item> remainingCourseItems = itemRepository.findAllByCourseIdAndIdNotIn(courseId, assignedItemIds);
+
+        return Stream.concat(
+                        assignedRewardMapper.assignedItemsToResponseDto(assignedItems).stream(),
+                        rewardMapper.itemsToEquipmentResponseDto(remainingCourseItems).stream()
+                )
+                .sorted(Comparator.comparing(response -> response.getBase().getOrderIndex()))
+                .toList();
+    }
+
+    public List<EquipmentChestResponseDto> getEquipmentChests(Long courseId) {
+        Long animalId = animalService.validateAndGetAnimalId(courseId);
+        List<AssignedChest> assignedChests = assignedRewardService.getAnimalAssignedChests(animalId);
+        List<Long> assignedChestIds = getAssignedRewardsIds(new ArrayList<>(assignedChests));
+        List<Chest> remainingCourseChests = chestRepository.findAllByCourseIdAndIdNotIn(courseId, assignedChestIds);
+
+        return Stream.concat(
+                        assignedRewardMapper.assignedChestsToResponseDto(courseId, animalId, assignedChests).stream(),
+                        rewardMapper.chestsToEquipmentResponseDto(remainingCourseChests).stream()
+                )
+                .sorted(Comparator.comparing(response -> response.getBase().getOrderIndex()))
+                .toList();
+    }
+
+    @Transactional
+    public void openChest(Long courseId, EquipmentChestOpenRequestDto requestDto) {
+        Long animalId = animalService.validateAndGetAnimalId(courseId);
+        AssignedChest assignedChest = assignedRewardService.getAssignedChestByIdAndAnimalId(
+                requestDto.getAssignedChestId(),
+                animalId
+        );
+
+        ZonedDateTime openDate = ZonedDateTime.now();
+        Chest chest = (Chest) Hibernate.unproxy(assignedChest.getReward());
+
+        List<AssignedItem> assignedItems = switch (chest.getBehavior()) {
+            case ONE_OF_MANY -> createAssignedItemsFromRequest(requestDto, assignedChest, openDate);
+            case ALL -> createAssignedItemsFromAssignedChest(chest, assignedChest, openDate);
+        };
+
+        assignedChest.setIsUsed(true);
+        assignedChest.setUsedDate(openDate);
+        assignedChest.getAssignedItems().addAll(assignedItems);
+
+        assignedRewardService.saveAssignedChest(assignedChest);
+        assignedRewardService.saveAssignedItems(assignedItems);
+        bonusXpCalculator.updateAnimalFlatBonusXp(animalId);
+
+    }
+
+
+    private List<Long> getAssignedRewardsIds(List<AssignedReward> assignedRewards) {
```

Comment:
```suggestion
    private List<Long> getAssignedRewardsIds(List<? extends AssignedReward> assignedRewards) {
```
I think that this would allow us to remove creating new arrays in `getEquipmentItems` and `getEquipmentChests`.